### PR TITLE
New Releases (Shanghai Support, EIP-4844 (experimental), Client Improvements)

### PR DESCRIPTION
**Pull request by holgerd77**
Follow-up on #2445 

Full round of client releases, releases are planned for next Tuesday, February 14 2023.

Release round includes full Shanghai support for the libraries, an experimental EIP-4844 Shard Blob Transactions implementation, various sync and UI/UX related client improvements following Interop week.

~WIP, submitted CHANGELOGs for selected libraries can be already reviewed.~

Ready for review. 🙂

### CODE REVIEW

**Comment by holgerd77** [comment]
Rebase this and fixed some tx README conflict.

File: packages/tx/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,72 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+  type: '0x05',
+  versionedHashes: ['0xabc...'], // Test with empty array on a first run
+  kzgCommitments: ['0xdef...'], // Test with empty array on a first run
+  blobs: ['0xghi...'], // Test with empty array on a first run
+}
+
+const tx = BlobEIP4844Transaction.fromTxData(txData, { common })
+```
+
+Note, the `versionedHashes`, `kzgCommitments`, and `blobs` are in reality 32 bytes or 4096 bytes in length but are trimmed here for brevity.
```
Comment (acolytec3) [inline packages/tx/CHANGELOG.md:71]: This should be clarified.  Only the blobs can be 4096 bytes.  The hashes and commitments are both 32 bytes.

File: packages/evm/CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,60 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+This precompile needs to have a working `kzg` library installation in the global namespace adhering to the [Kzg](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/tx/src/depInterfaces.ts) interface defined in the `@ethereumjs/tx` library.
+
+This release has been tested with the [c-kzg](https://github.com/ethereum/c-kzg-4844) library which can be installed with `npm install c-kzg`.
+
+This library then needs to be imported along the other library imports:
+
+```typescript
+import { Common, Hardfork } from '@ethereumjs/common'
+import * as kzg from 'c-kzg'
+import { EVM } from '@ethereumjs/evm'
```
Comment (acolytec3) [inline packages/evm/CHANGELOG.md:56]: Maybe also add a note that kzg needs to be initialized using
`initKzg(kzg, 'path/to/my/trusted_setup.txt')`.

**Comment by holgerd77** [comment]
This is now complete, maybe a bit rough still since I pushed relatively quickly. Nevertheless open for a first round of full review. 🙂

File: packages/trie/CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 (modification: no type change headlines) and this project adheres to
 [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## 5.0.3 - 2023-02-16
+
```
Comment (acolytec3) [inline packages/trie/CHANGELOG.md:10]: Do we need to mention the readme updates from #2525.  I can't remember if you normally include doc changes within the Changelog or not (since it's not a functional change to the code).

File: packages/tx/CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,72 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+##### Usage
+
+See the following code snipped for an example on how to instantiate (using the `c-kzg` module for our KZG dependency).
+
+```typescript
+import { Chain, Common, Hardfork } from '@ethereumjs/common'
+import { BlobEIP4844Transaction, initKzg } from '@ethereumjs/tx'
+import * as kzg from 'c-kzg'
+
+initKzg(kzg, 'path/to/my/trusted_setup.txt')
```
Comment (acolytec3) [inline packages/tx/CHANGELOG.md:45]: We should include this line plus the imports for `kzg` and `initKzg` in `block`, `evm,` and `blockchain` as well.

**Review by acolytec3 — CHANGES_REQUESTED** [review]
Looks good so far.  Just need to add the `initKzg` line to the usage examples for 4844 in the impacted modules.

File: packages/evm/CHANGELOG.md (hunk 2 above)
Comment (holgerd77) [reply packages/evm/CHANGELOG.md]: I've now consolidated the KZG instructions within a single source of truths, see 3f44711297c6d3719f4bb4028f7b0b2567412ad4.

File: packages/trie/CHANGELOG.md (hunk 3 above)
Comment (holgerd77) [reply packages/trie/CHANGELOG.md:10]: I guess that's not so important, will skip, these small changes will just make its way into the code docs organically and hopefully help users a bit over time. There are no absolute "rules" what to include or not to include in the CHANGELOGs, these should just provide a helpful overview on what's new and relevant and might need some attention.

File: packages/tx/CHANGELOG.md (hunk 4 above)
Comment (holgerd77) [reply packages/tx/CHANGELOG.md:38]: Ah `blockchain`, yes you are right, forgot about that. Even more useful to have these consolidated instructions now, this would otherwise get super redundant. Have added (not pushed yet).

**Comment by holgerd77** [comment]
Ok, this is now ready with an extensive Client CHANGELOG also added. 🙂 

It would be good if this would get a deeper round of reviews soon, so that we are ready on Tuesday, this the scope of this PR is pretty extensive. It would be good if all CHANGELOG parts are re-read at least once by someone, also cross-checks on completeness and correctness of version bumps would be useful.

**Review by acolytec3 — CHANGES_REQUESTED** [review]
This looks good to me, with one small edit (I think that's what was intended but will let you confirm).

File: packages/block/CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,42 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+const common = new Common({ chain: Chain.Mainnet, hardfork: Hardfork.Shanghai })
+```
+
+### EIP-4844 Shard Blob Transactions Support (experimental)
+
+This release supports an experimental version of the blob transaction type introduced with [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844) as being specified in the [01d3209](https://github.com/ethereum/EIPs/commit/01d320998d1d53d95f347b5f43feaf606f230703) EIP version from February 8, 2023 and deployed along `eip4844-devnet-4` (January 2023).
+
+#### Initialization
+
+To create blocks which include blob transactions you have to active EIP-4844 in the associated `@ethereumjs/common` library:
```
Comment (gabrocheleau) [inline packages/block/CHANGELOG.md:29]:
```suggestion
To create blocks which include blob transactions you have to activate EIP-4844 in the associated `@ethereumjs/common` library:
```

File: packages/block/CHANGELOG.md [hunk 6]
```diff
@@ -6,6 +6,42 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+
+```typescript
+import { Common, Chain, Hardfork } from '@ethereumjs/common'
+
+const common = new Common({ chain: Chain.Mainnet, hardfork: Hardfork.Shanghai, eips: [4844] })
+```
+
+**Note:** Working with blob transactions needs a manual KZG library installation and global initialization, see [KZG Setup](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/tx/README.md#kzg-setup) for instructions.
```
Comment (gabrocheleau) [inline packages/block/CHANGELOG.md:37]:
```suggestion
**Note:** Working with blob transactions requires a manual KZG library installation and global initialization, see [KZG Setup](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/tx/README.md#kzg-setup) for instructions.
```

File: packages/blockchain/CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,46 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+
+- Blockchain class: `async getBlock(blockId: Buffer | number | bigint): Promise<Block>`
+- Blockchain interface: `getBlock(blockId: Buffer | number | bigint): Promise<Block | null>`
+
+So the Blockchain interface was - falsely - claiming that there would be the possibility of a `null` value returned in the case of a block not being found while the actual implementation was throwing an error in such a case.
+
+We now fixed this by removing the `null` from the interface return values - see PR [#2524](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2524), after exploring the other way around as well (and the reverting), see PR [#2516](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2516).
+
+While this might lead to breaking code constallations on the TypeScript level if this `null` value is picked up we felt this is the right thing to do since this divergence would otherwise continue to "trick" people into assuming and dealing with `null` values for non-existing-block assumptions in their code and continue to produce eventual bugs (we actually fell over this ourselves).
```
Comment (gabrocheleau) [inline packages/blockchain/CHANGELOG.md:40]:
```suggestion
While this might lead to breaking code installations on the TypeScript level if this `null` value is picked up we felt this is the right thing to do since this divergence would otherwise continue to "trick" people into assuming and dealing with `null` values for non-existing-block assumptions in their code and continue to produce eventual bugs (we actually fell over this ourselves).
```
Not sure what `constallations` means, did yo mean installation?

File: packages/client/CHANGELOG.md [hunk 8]
```diff
@@ -6,6 +6,75 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+See Interop EIP-4844 Tracking Issue [#2494](https://github.com/ethereumjs/ethereumjs-monorepo/issues/2494) for information on how to run the client in the context of an EIP-4844 testnet.
+
+#### Related Changes
+
+- Add checks for replacement data gas too low for blob txs, PR [#2503](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2503)
+- 4844 Engine API Fixes, PR [#2508](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2508)
+
+### Post-Merge UX Improvements
+
+The post-Merge client log flow where the EthereumJS client serves as a responsive execution client in combination with a consensus layer client like Lodestar has been completely overhauled, see PRs [#2497](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2497), [#2506](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2506) and [#2510](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2497).
```
Comment (gabrocheleau) [inline packages/client/CHANGELOG.md:44]:
```suggestion
The post-Merge client log flow where the EthereumJS client serves as a responsive execution client in combination with a consensus layer client like Lodestar has been completely overhauled, see PRs [#2497](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2497), [#2506](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2506) and [#2510](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2510).
```

File: packages/tx/README.md [hunk 9]
```diff
@@ -112,15 +134,15 @@ const txData = {
   chainId: '0x01',
   accessList: [],
   type: '0x05',
-  versionedHashes: ['0xabc...'],
-  kzgCommitments: ['0xdef...'],
-  blobs: ['0xghi...'],
+  versionedHashes: ['0xabc...'], // Test with empty array on a first run
+  kzgCommitments: ['0xdef...'], // Test with empty array on a first run
+  blobs: ['0xghi...'], // Test with empty array on a first run
```
Comment (gabrocheleau) [inline packages/tx/README.md:139]: Could we elaborate a bit more on why someone would "test with empty array on a first run"? (Perhaps in the instructions)

File: packages/vm/README.md [hunk 10]
```diff
@@ -170,6 +204,24 @@ const vm = new VM({ common })
 
 For a list with supported EIPs see the [@ethereumjs/evm](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/evm) documentation.
 
+### EIP-4844 Shard Blob Transactions Support (experimental)
+
+This library supports an experimental version of the blob transaction type introduced with [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844) as being specified in the [01d3209](https://github.com/ethereum/EIPs/commit/01d320998d1d53d95f347b5f43feaf606f230703) EIP version from February 8, 2023 and deployed along `eip4844-devnet-4` (January 2023) starting with `v1.3.0`.
+
+#### Initialization
+
+To run VM/EVM related EIP-4844 functionality you have to active the EIP in the associated `@ethereumjs/common` library:
```
Comment (gabrocheleau) [inline packages/vm/README.md:213]:
```suggestion
To run VM/EVM related EIP-4844 functionality you have to activate the EIP in the associated `@ethereumjs/common` library:
```

File: packages/vm/CHANGELOG.md [hunk 11]
```diff
@@ -6,6 +6,54 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+
+Note: that this is only a finalizing release by e.g. integrating an updated `@ethereumjs/common` library with an updated Shanghai HF setting and all Shanghai related EIP functionality has been already released in former releases. Do a fulltext search on the EIP numbers in the EVM/VM CHANGELOG files for additional information and usage instructions.
+
+### Experimental EIP-4844 Shard Blob Transactions Support
+
+This release supports an experimental version of the blob transaction type introduced with [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844) as being specified in the [01d3209](https://github.com/ethereum/EIPs/commit/01d320998d1d53d95f347b5f43feaf606f230703) EIP version from February 8, 2023 and deployed along `eip4844-devnet-4` (January 2023), see PR [#2349](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2349) as well as PRs [#2522](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2522) and [#2526](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2526).
+
+#### Initialization
+
+To run VM/EVM related EIP-4844 functionality you have to active the EIP in the associated `@ethereumjs/common` library:
```
Comment (gabrocheleau) [inline packages/vm/CHANGELOG.md:31]:
```suggestion
To run VM/EVM related EIP-4844 functionality you have to activate the EIP in the associated `@ethereumjs/common` library:
```

File: packages/util/CHANGELOG.md [hunk 12]
```diff
@@ -6,6 +6,14 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 (modification: no type change headlines) and this project adheres to
 [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## 8.0.4 - 2023-02-21
+
+- Removed `async` library dependency, PR [#2514](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2514)
+- New `GWEI_TO_WEI` constant in a newly created `units` module, PR [#2483](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2483)
+- Change withdrawal amount representation from Wei to Gwei (see EIP-4895 PR [#6325](https://github.com/ethereum/EIPs/pull/6325)) in `withdrawal` module `Withdrawal` class, PR [#2483](Change withdrawal amount representation from Wei to Gwei
+  )
```
Comment (gabrocheleau) [inline packages/util/CHANGELOG.md:14]:
```suggestion
- Change withdrawal amount representation from Wei to Gwei (see EIP-4895 PR [#6325](https://github.com/ethereum/EIPs/pull/6325)) in `withdrawal` module `Withdrawal` class, PR [#2483](Change withdrawal amount representation from Wei to Gwei)
```

**Review by gabrocheleau — COMMENTED** [review]
Minor corrections :) looks good!

File: packages/blockchain/CHANGELOG.md (hunk 7 above)
Comment (holgerd77) [reply packages/blockchain/CHANGELOG.md:40]: Ah, not, this is constallations in the sense that some people might use this code in a constallation (in the sense of combination / combinatory setup) that this might break their code. I am ok with `installations` as well though, especially since I mistyped `constellation`. 😋

File: packages/tx/README.md (hunk 9 above)
Comment (holgerd77) [reply packages/tx/README.md:139]: This actually has no magical subcontext, but is rather to see if things work in general, which is easier to test/split without the actual data (then one just see if the KGZ setup works fine, the block constructs without errors,...).

**Comment by holgerd77** [comment]
@gabrocheleau thanks for reviewing and the direct fixes! 🙏 🙂

File: packages/util/CHANGELOG.md [hunk 13]
```diff
@@ -6,6 +6,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 (modification: no type change headlines) and this project adheres to
 [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## 8.0.4 - 2023-02-21
+
+- Removed `async` library dependency, PR [#2514](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2514)
+- New `GWEI_TO_WEI` constant in a newly created `units` module, PR [#2483](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2483)
+- Change withdrawal amount representation from Wei to Gwei (see EIP-4895 PR [#6325](https://github.com/ethereum/EIPs/pull/6325)) in `withdrawal` module `Withdrawal` class, PR [#2483](Change withdrawal amount representation from Wei to Gwei
```
Comment (gabrocheleau) [inline packages/util/CHANGELOG.md:13]:
```suggestion
- Change withdrawal amount representation from Wei to Gwei (see EIP-4895 PR [#6325](https://github.com/ethereum/EIPs/pull/6325)) in `withdrawal` module `Withdrawal` class, PR [#2483](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2483)
```

File: packages/util/CHANGELOG.md [hunk 14]
```diff
@@ -6,6 +6,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 (modification: no type change headlines) and this project adheres to
 [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## 8.0.4 - 2023-02-21
+
+- Removed `async` library dependency, PR [#2514](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2514)
+- New `GWEI_TO_WEI` constant in a newly created `units` module, PR [#2483](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2483)
+- Change withdrawal amount representation from Wei to Gwei (see EIP-4895 PR [#6325](https://github.com/ethereum/EIPs/pull/6325)) in `withdrawal` module `Withdrawal` class, PR [#2483](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2483)
```
Comment (gabrocheleau) [inline packages/util/CHANGELOG.md:13]: I updated #2483 to link the PR number but just noticed it's already refered to above. Could we actually merge the two points into one?

File: packages/client/CHANGELOG.md [hunk 15]
```diff
@@ -6,6 +6,78 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
```
Comment (gabrocheleau) [inline packages/client/CHANGELOG.md:45]:
```suggestion
The post-Merge client log flow where the EthereumJS client serves as a responsive execution client in combination with a consensus layer client like Lodestar has been completely overhauled, see PRs [#2497](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2497), [#2506](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2506) and [#2510](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2510).
```

**Review by gabrocheleau — APPROVED** [review]
Approving. left one small comment.

**Comment by holgerd77** [comment]
@gabrocheleau thanks for the continued reviews, sorry sorry, I just realized that I force-pushed your other updates away yesterday accidentally. 😳

I will nevertheless continue with the releases, this would otherwise hold up the process too much. Maybe we can still add in a follow up PR.

Again, sorry, was a bit tired.

**Review by g11tech — COMMENTED** [review]
all good :+1:
